Cast two-digit version numbers to character

peterhurford opened the pull request:
Resolves https://github.com/robertzk/lockbox/issues/119

peterhurford committed:
tests

peterhurford committed:
Versioning

peterhurford commented:
@russellpierce

peterhurford committed:
rm test.yml

kirillseva commented on R/lockbox.R:
```diff
@@ -79,6 +72,17 @@ lockbox.default <- function(obj) {
+  } else {
+    lock$packages[vapply(lock$packages, `[[`, character(1), "name") %in% lock[[env]]]
+  }
+  lapply(lock, function(xs) lapply(xs, function(x) as.character(x)))
```
`lapply(lock, function(xs) lapply(xs, as.character))`

kirillseva replied:
hopefully booleans will keep being parsed / used correctly

peterhurford replied:
We can add some tests for that, but I'll have to do that later.

kirillseva commented:
foo.yml
```yaml
name: RSQLite
version: 2.0
load: false
```

in R:
```r
kk <- yaml::yaml.load_file('foo.yml')
kk$version %>% as.character
kk
# $name
# [1] "RSQLite"
#
# $version
# [1] 2
#
# $load
# [1] FALSE

kk$version %>% as.character
# [1] "2"
```

peterhurford committed:
fix env default on parse_lock

kirillseva commented on tests/testthat/test-parse-lock.R:
```diff
@@ -0,0 +1,64 @@
+      }
+    }
+  })
+})
```
these tests don't take issues like #119 into account, since it's the confusion happening in yaml parsing. Consider adding some dummy lockfiles in `inst/` and doing `yaml::yaml.load_file` on those, in addition to parsing data that's already in R in a correct format.

peterhurford replied:
On it.

peterhurford committed:
test some actual yml

peterhurford commented:
@kirillseva I see what you mean about how `1.0` gets turned into `"1"`...

kirillseva commented on tests/testthat/test-parse-lock.R:
```diff
@@ -0,0 +1,114 @@
+      }
+    }
+  })
+})
```
👍 x 💯

peterhurford committed:
use custom float handler in yaml

peterhurford commented:
@kirillseva Looks like a custom float handler solves the issue!

peterhurford committed:
simpler nested lapply

peterhurford commented:
Please test this more :)

kirillseva commented on R/lockbox.R:
```diff
@@ -54,7 +54,10 @@ lockbox <- function(file_or_list, env = getOption("lockbox.env", "!packages")) {
 parse_lock <- function(lock, env = getOption("lockbox.env", "!packages")) {
-  if (is.character(lock) && length(lock) == 1) { lock <- yaml::yaml.load_file(lock) }
+  if (is.character(lock) && length(lock) == 1) {
+    handle_float_version <- function(x) { sprintf("%.1f", 1.0) }
```
s/1.0/x

kirillseva commented on R/lockbox.R:
```diff
@@ -70,13 +52,24 @@ lockbox.list <- function(lock, env) {
+
+parse_lock <- function(lock, env = getOption("lockbox.env", "!packages")) {
+  if (is.character(lock) && length(lock) == 1) {
+    handle_float_version <- function(x) { sprintf("%.1f", 1.0) }
```
s/1.0/x

kirillseva commented:
I like the tests and the improved modularity of this pull-request, but I feel like introducing custom yaml loading rules and casting everything as characters, including booleans, is overengineering the problem and makes debugging future issues more complicated.

Especially when the solution to the problem is using a couple of quotes here and there

kirillseva committed:
Update lockbox.R

russellpierce commented:
I like the idea that this would make the lockfile just work and hide the complexity from the user. 

What if someone decides to version their package just by integers with no decimal, would this still work?

What gets me with 119 is that it is just detecting the existing version that fails and triggers a reinstall. Lockbox then goes on to load things just fine.

If a solution like this isn't accepted, could we find a middle ground with an informative warning when the version specified risks running into this issue?

kirillseva commented:
Or we can update the readme so that all examples show version listed in
quotes and avoid overengineering
On Sun, Aug 20, 2017 at 3:11 PM Russell S. Pierce <notifications@github.com>
wrote:

> I like the idea that this would make the lockfile just work and hide the
> complexity from the user.
>
> What if someone decides to version their package just by integers with no
> decimal, would this still work?
>
> What gets me with 119 is that it is just detecting the existing version
> that fails and triggers a reinstall. Lockbox then goes on to load things
> just fine.
>
> If a solution like this isn't accepted, could we find a middle ground with
> an informative warning when the version specified risks running into this
> issue?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/robertzk/lockbox/pull/120#issuecomment-323605227>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AEO9uLsUM4WYCakVzQ_D_UCPWC46kAY1ks5saITOgaJpZM4O8biA>
> .
>

peterhurford committed:
fix logical handling and add more tests

peterhurford commented:
> I like the idea that this would make the lockfile just work and hide the complexity from the user.

> is overengineering the problem and makes debugging future issues more complicated

It's definitely a trade-off. I'm open for a vote on this.

~

> I like the tests and the improved modularity of this pull-request

Either way, I do think we should keep that. I can resubmit the PR if desired.

~

> casting everything as characters, including booleans

Good catch. I fixed that so that booleans stay as booleans. I added test coverage for that too.

And thanks for catching the embarrassing `sprintf("%.1f", 1.0)` fix. ...This is why I made quickcheck for R. ...I added some more tests around that.

~

> What if someone decides to version their package just by integers with no decimal, would this still work?

Yes.

russellpierce commented:
Vote: Yes

The PR is well compartmentalized and tested.  For now it solves a problem that affects the usability of the package.  I grant that the PR represents extra complexity.  However, there is nothing to say we can't revert that complexity down the line if it becomes a problem.  In the near term I don't see much reason to shy away from it.

At the very least, let's get the passing tests merged - then we can leave updating the docs or adding a warning to another PR.

kirillseva commented:
Oh tests are here to stay, I'm not letting them go :) but I am of an
opinion that we do not need any extra magical typecasting under the hood
On Tue, Aug 22, 2017 at 2:45 PM Russell S. Pierce <notifications@github.com>
wrote:

> Vote: Yes
>
> The PR is well compartmentalized and tested. For now it solves a problem
> that affects the usability of the package. I grant that the PR represents
> extra complexity. However, there is nothing to say we can't revert that
> complexity down the line if it becomes a problem. In the near term I don't
> see much reason to shy away from it.
>
> At the very least, let's get the passing tests merged - then we can leave
> updating the docs or adding a warning to another PR.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/robertzk/lockbox/pull/120#issuecomment-324131760>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AEO9uFAtUdw7WpJw1kkyGlnvJxJ1-A6dks5say_vgaJpZM4O8biA>
> .
>

kirillseva commented:
Here's how I would deal with this situation if it were `kirillseva/lockbox` :)

1. Merge https://github.com/robertzk/lockbox/pull/121
2. Pluck tests out of this PR and merge them
3. Do not do any hidden logic in between yaml <--> R reading as that could potentially cause a lot of confusion at the most uncomfortable time

But that's just like my opinion, we're all obviously waiting on @robertzk's thoughts on this matter

peterhurford commented:
@kirillseva I'm happy to let @robertzk decide. I can assist with making a new PR with the tests + refactor if @robertzk decides to go off of #121 instead of #120.

peterhurford commented:
Either @robertzk merged https://github.com/robertzk/lockbox/pull/121 without thinking of this PR or he has voted against the custom character parsing. @robertzk can you weigh in explicitly?

peterhurford commented:
@robertzk ping

robertzk commented on R/lockbox.R:
```diff
@@ -70,13 +52,30 @@ lockbox.list <- function(lock, env) {
+parse_lock <- function(lock, env = getOption("lockbox.env", "!packages")) {
+  if (is.character(lock) && length(lock) == 1) {
+    handle_float_version <- function(x) {
+      if (is.numeric(x)) { sprintf("%.1f", x) }
```
Won't this fail on versions like `0.11`? I think we just want `as.character(x)` or a manual strsplit on the `.` if it exists.

```r
R-ENGINES> sprintf("%.1f", 0.11)
[1] "0.1"
```

robertzk replied:
Wouldn't `x` always be numeric given that it's the float yml handler?

russellpierce replied:
You mentioned strsplit on `.` would we also need it for `-`?  Is there actually a limited character set that CRAN enforces for versions?

robertzk replied:
@russellpierce This logic should only apply to yml values detected as being float values, e.g.,

```yml
version: 1.0
```

rather than

```yml
version: "1.0"
```

robertzk commented on R/lockbox.R:
```diff
@@ -70,13 +52,30 @@ lockbox.list <- function(lock, env) {
+       "instead I got a ", sQuote(class(lock)[1]), ".")
+  }
+  if (is.null(lock$packages)) stop("Invalid config. Make sure your config format is correct")
+  lock <- if (identical(env, "!packages") || is.null(lock[[env]])) {
```
The style guide preference here is to assign `lock` in the body for readability.

robertzk commented on tests/testthat/data/test-envs.yml:
```diff
@@ -0,0 +1,15 @@
+packages:
```
Let's add a `0.999` test here as well per the above point.

robertzk reviewed (changes requested):
Just a minor comment on versions like `0.x` where `x > 9`. Otherwise looks good!

robertzk commented:
@peterhurford Ping!

robertzk committed:
Remove extraneous float handler (identity is sufficient).

robertzk commented:
@peterhurford It looks like the float handler [wasn't doing](https://github.com/robertzk/lockbox/pull/120/commits/afa12f8dcb0433eb665843e3461a70c65b7ec34f) as much as expected. (Putting a `stop` in the `is.numeric(x)` branch showed it doesn't execute at all during tests.) I'll merge when green.

peterhurford commented:
@robertzk Thanks for finishing this up!

peterhurford commented:
@robertzk @kirillseva Should we revert https://github.com/robertzk/lockbox/pull/121 in light of this PR merge?

robertzk commented:
@peterhurford [Good point!](https://github.com/robertzk/lockbox/pull/127)
